File: go/internal/hw/hw.go
package hw

import (
	"encoding/json"
	"strconv"
	"strings"

	"github.com/marknorgren/about-this-mac/go/internal/execx"
)

type Memory struct {
	Total        string `json:"total"`
	Type         string `json:"type"`
	Speed        string `json:"speed"`
	Manufacturer string `json:"manufacturer"`
	ECC          bool   `json:"ecc"`
}

type Storage struct {
	Name        string `json:"name"`
	Model       string `json:"model"`
	Revision    string `json:"revision"`
	Serial      string `json:"serial"`
	Size        string `json:"size"`
	Type        string `json:"type"`
	Protocol    string `json:"protocol"`
	Trim        bool   `json:"trim"`
	SmartStatus string `json:"smart_status"`
	Removable   bool   `json:"removable"`
	Internal    bool   `json:"internal"`
}

type Info struct {
	ModelName         string              `json:"model_name"`
	DeviceIdentifier  string              `json:"device_identifier"`
	ModelNumber       string              `json:"model_number"`
	SerialNumber      string              `json:"serial_number"`
	Processor         string              `json:"processor"`
	CPUCores          int                 `json:"cpu_cores"`
	PerformanceCores  int                 `json:"performance_cores"`
	EfficiencyCores   int                 `json:"efficiency_cores"`
	GPUCores          int                 `json:"gpu_cores"`
	Memory            Memory              `json:"memory"`
	Storage           Storage             `json:"storage"`
	Graphics          []map[string]string `json:"graphics"`
	BluetoothChipset  string              `json:"bluetooth_chipset"`
	BluetoothFirmware string              `json:"bluetooth_firmware"`
	BluetoothTrans    string              `json:"bluetooth_transport"`
	MacOSVersion      string              `json:"macos_version"`
	MacOSBuild        string              `json:"macos_build"`
	Uptime            *int64              `json:"uptime"`
	ReleaseDate       string              `json:"release_date"`
	ModelSize         string              `json:"model_size"`
	ModelYear         string              `json:"model_year"`
}

type spHardware struct {
	SPHardwareDataType []map[string]any `json:"SPHardwareDataType"`
}

type spDisplays struct {
	SPDisplaysDataType []struct {
		Displays []struct {
			Pixels     string `json:"_spdisplays_pixels"`
			Connection string `json:"spdisplays_connection_type"`
		} `json:"spdisplays_ndrvs"`
	} `json:"SPDisplaysDataType"`
}

// Gather collects hardware info via the provided runner. This is a phase-1
// stub: it populates the bare minimum needed for --format simple. Subsequent
// patches will fill in the rest field-by-field, with byte-parity tests against
// Python output.
func Gather(r execx.Runner) (Info, error) {
	var info Info

	out, _, err := r.Run("system_profiler", "SPHardwareDataType", "-json")
	if err == nil && out != "" {
		var parsed spHardware
		if jerr := json.Unmarshal([]byte(out), &parsed); jerr == nil && len(parsed.SPHardwareDataType) > 0 {
			hw := parsed.SPHardwareDataType[0]
			info.ModelName = asString(hw["machine_name"])
			info.DeviceIdentifier = asString(hw["machine_model"])
			info.ModelNumber = asString(hw["model_number"])
			info.SerialNumber = asString(hw["serial_number"])
			info.Processor = firstString(hw, "chip_info", "chip_type", "cpu_type")
			applyMarketingName(&info, info.ModelName)
		}
	}

	if mem, _, _ := r.Run("sysctl", "-n", "hw.memsize"); mem != "" {
		info.Memory.Total = memoryGB(strings.TrimSpace(mem))
	}
	if info.Memory.Total == "" && out != "" {
		var parsed spHardware
		if jerr := json.Unmarshal([]byte(out), &parsed); jerr == nil && len(parsed.SPHardwareDataType) > 0 {
			info.Memory.Total = normalizeMemoryTotal(asString(parsed.SPHardwareDataType[0]["physical_memory"]))
		}
	}
	if info.ModelSize == "" {
		info.ModelSize = modelSizeFromDisplays(r)
	}
	if info.ReleaseDate == "" {
		info.ReleaseDate = releaseDateFromChip(info.Processor)
	}
	if info.ModelYear == "" {
		info.ModelYear = yearFromReleaseDate(info.ReleaseDate)
	}

	if ver, _, _ := r.Run("sw_vers", "-productVersion"); ver != "" {
		info.MacOSVersion = strings.TrimSpace(ver)
	}

	return info, nil
}

func firstString(values map[string]any, keys ...string) string {
	for _, key := range keys {
		if value := asString(values[key]); value != "" {
			return value
		}
	}
	return ""
}

func asString(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

func memoryGB(raw string) string {
	bytes, err := strconv.ParseInt(strings.TrimSpace(raw), 10, 64)
	if err != nil || bytes <= 0 {
		return ""
	}
	return strconv.FormatInt(bytes/(1024*1024*1024), 10) + "GB"
}

func normalizeMemoryTotal(raw string) string {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return ""
	}
	return strings.ReplaceAll(raw, " GB", "GB")
}

func applyMarketingName(info *Info, marketingName string) {
	if marketingName == "" {
		return
	}
	if before, _, ok := strings.Cut(marketingName, "("); ok {
		if name := strings.TrimSpace(before); name != "" {
			info.ModelName = name
		}
	}
	if size := modelSizeFromMarketingName(marketingName); size != "" {
		info.ModelSize = size
	}
	if year := firstYear(marketingName); year != "" {
		info.ModelYear = year
	}
}

func modelSizeFromMarketingName(name string) string {
	idx := strings.Index(name, "-inch")
	if idx == -1 {
		return ""
	}
	start := idx
	for start > 0 && name[start-1] >= '0' && name[start-1] <= '9' {
		start--
	}
	if start == idx {
		return ""
	}
	return name[start : idx+len("-inch")]
}

func modelSizeFromDisplays(r execx.Runner) string {
	out, _, err := r.Run("system_profiler", "SPDisplaysDataType", "-json")
	if err != nil || out == "" {
		return ""
	}
	var parsed spDisplays
	if err := json.Unmarshal([]byte(out), &parsed); err != nil {
		return ""
	}
	for _, card := range parsed.SPDisplaysDataType {
		for _, display := range card.Displays {
			if !strings.Contains(strings.ToLower(display.Connection), "internal") {
				continue
			}
			if size := screenSizeFromResolution(display.Pixels); size != "" {
				return size
			}
		}
	}
	return ""
}

func screenSizeFromResolution(resolution string) string {
	parts := strings.Split(strings.ToLower(resolution), "x")
	if len(parts) == 0 {
		return ""
	}
	width, err := strconv.Atoi(strings.TrimSpace(parts[0]))
	if err != nil {
		return ""
	}
	switch width {
	case 3456:
		return "16-inch"
	case 3024:
		return "14-inch"
	case 2560, 1440, 1680:
		return "13-inch"
	case 2880, 1920:
		return "15-inch"
	case 2304:
		return "12-inch"
	default:
		return ""
	}
}

func releaseDateFromChip(chip string) string {
	for _, candidate := range []struct {
		token string
		date  string
	}{
		{"M4", "Mar 2024"},
		{"M3", "Oct 2023"},
		{"M2", "Jan 2023"},
		{"M1", "Nov 2020"},
	} {
		if strings.Contains(chip, candidate.token) {
			return candidate.date
		}
	}
	return ""
}

func yearFromReleaseDate(date string) string {
	return firstYear(date)
}

func firstYear(s string) string {
	fields := strings.FieldsFunc(s, func(r rune) bool { return r < '0' || r > '9' })
	for _, field := range fields {
		if len(field) == 4 && strings.HasPrefix(field, "20") {
			return field
		}
	}
	return ""
}

File: rust/src/hw.rs
use crate::execx::Runner;
use serde::{Deserialize, Serialize};

#[derive(Default, Debug, Serialize, Deserialize)]
pub struct Memory {
    pub total: String,
    pub r#type: String,
    pub speed: String,
    pub manufacturer: String,
    pub ecc: bool,
}

#[derive(Default, Debug, Serialize, Deserialize)]
pub struct Storage {
    pub name: String,
    pub model: String,
    pub revision: String,
    pub serial: String,
    pub size: String,
    pub r#type: String,
    pub protocol: String,
    pub trim: bool,
    pub smart_status: String,
    pub removable: bool,
    pub internal: bool,
}

#[derive(Default, Debug, Serialize, Deserialize)]
pub struct Info {
    pub model_name: String,
    pub device_identifier: String,
    pub model_number: String,
    pub serial_number: String,
    pub processor: String,
    pub cpu_cores: u32,
    pub performance_cores: u32,
    pub efficiency_cores: u32,
    pub gpu_cores: u32,
    pub memory: Memory,
    pub storage: Storage,
    pub graphics: Vec<std::collections::BTreeMap<String, String>>,
    pub bluetooth_chipset: String,
    pub bluetooth_firmware: String,
    pub bluetooth_transport: String,
    pub macos_version: String,
    pub macos_build: String,
    pub uptime: Option<i64>,
    pub release_date: String,
    pub model_size: String,
    pub model_year: String,
}

#[derive(Deserialize)]
struct SpHardware {
    #[serde(rename = "SPHardwareDataType")]
    items: Vec<serde_json::Value>,
}

/// Gather collects hardware info via the provided runner. Phase 1 stub:
/// populates only what `--format simple` needs. Will be extended field by
/// field, byte-parity tested against Python output.
pub fn gather(runner: &dyn Runner) -> std::io::Result<Info> {
    let mut info = Info::default();

    let (out, _code) = runner.run("system_profiler", &["SPHardwareDataType", "-json"])?;
    if !out.is_empty() {
        if let Ok(parsed) = serde_json::from_str::<SpHardware>(&out) {
            if let Some(hw) = parsed.items.first() {
                info.model_name = json_str(hw, "machine_name");
                info.device_identifier = json_str(hw, "machine_model");
                info.model_number = json_str(hw, "model_number");
                info.serial_number = json_str(hw, "serial_number");
                info.processor = first_json_str(hw, &["chip_info", "chip_type", "cpu_type"]);
                apply_marketing_name(&mut info, &json_str(hw, "machine_name"));
                info.memory.total = normalize_memory_total(&json_str(hw, "physical_memory"));
            }
        }
    }

    let (memsize, _) = runner.run("sysctl", &["-n", "hw.memsize"])?;
    if let Some(total) = memory_gb(memsize.trim()) {
        info.memory.total = total;
    }
    if info.model_size.is_empty() {
        info.model_size = model_size_from_displays(runner)?;
    }
    if info.release_date.is_empty() {
        info.release_date = release_date_from_chip(&info.processor);
    }
    if info.model_year.is_empty() {
        info.model_year = first_year(&info.release_date);
    }

    let (ver, _) = runner.run("sw_vers", &["-productVersion"])?;
    if !ver.is_empty() {
        info.macos_version = ver.trim().to_string();
    }

    Ok(info)
}

fn json_str(v: &serde_json::Value, key: &str) -> String {
    v.get(key)
        .and_then(|x| x.as_str())
        .unwrap_or_default()
        .to_string()
}

fn first_json_str(v: &serde_json::Value, keys: &[&str]) -> String {
    for key in keys {
        let value = json_str(v, key);
        if !value.is_empty() {
            return value;
        }
    }
    String::new()
}

fn memory_gb(raw: &str) -> Option<String> {
    let bytes = raw.parse::<u64>().ok()?;
    if bytes == 0 {
        return None;
    }
    Some(format!("{}GB", bytes / 1024 / 1024 / 1024))
}

fn normalize_memory_total(raw: &str) -> String {
    raw.trim().replace(" GB", "GB")
}

fn apply_marketing_name(info: &mut Info, marketing_name: &str) {
    if marketing_name.is_empty() {
        return;
    }
    if let Some((name, _)) = marketing_name.split_once('(') {
        let name = name.trim();
        if !name.is_empty() {
            info.model_name = name.to_string();
        }
    }
    let size = model_size_from_marketing_name(marketing_name);
    if !size.is_empty() {
        info.model_size = size;
    }
    let year = first_year(marketing_name);
    if !year.is_empty() {
        info.model_year = year;
    }
}

fn model_size_from_marketing_name(name: &str) -> String {
    let Some(idx) = name.find("-inch") else {
        return String::new();
    };
    let bytes = name.as_bytes();
    let mut start = idx;
    while start > 0 && bytes[start - 1].is_ascii_digit() {
        start -= 1;
    }
    if start == idx {
        return String::new();
    }
    name[start..idx + "-inch".len()].to_string()
}

fn model_size_from_displays(runner: &dyn Runner) -> std::io::Result<String> {
    let (out, _) = runner.run("system_profiler", &["SPDisplaysDataType", "-json"])?;
    if out.is_empty() {
        return Ok(String::new());
    }
    let Ok(parsed) = serde_json::from_str::<serde_json::Value>(&out) else {
        return Ok(String::new());
    };
    let Some(cards) = parsed
        .get("SPDisplaysDataType")
        .and_then(serde_json::Value::as_array)
    else {
        return Ok(String::new());
    };
    for card in cards {
        let Some(displays) = card
            .get("spdisplays_ndrvs")
            .and_then(serde_json::Value::as_array)
        else {
            continue;
        };
        for display in displays {
            let connection = json_str(display, "spdisplays_connection_type").to_lowercase();
            if !connection.contains("internal") {
                continue;
            }
            let size = screen_size_from_resolution(&json_str(display, "_spdisplays_pixels"));
            if !size.is_empty() {
                return Ok(size);
            }
        }
    }
    Ok(String::new())
}

fn screen_size_from_resolution(resolution: &str) -> String {
    let normalized = resolution.to_lowercase();
    let Some((width, _)) = normalized.split_once('x') else {
        return String::new();
    };
    match width.trim().parse::<u32>() {
        Ok(3456) => "16-inch".to_string(),
        Ok(3024) => "14-inch".to_string(),
        Ok(2560 | 1440 | 1680) => "13-inch".to_string(),
        Ok(2880 | 1920) => "15-inch".to_string(),
        Ok(2304) => "12-inch".to_string(),
        _ => String::new(),
    }
}

fn release_date_from_chip(chip: &str) -> String {
    for (token, date) in [
        ("M4", "Mar 2024"),
        ("M3", "Oct 2023"),
        ("M2", "Jan 2023"),
        ("M1", "Nov 2020"),
    ] {
        if chip.contains(token) {
            return date.to_string();
        }
    }
    String::new()
}

fn first_year(value: &str) -> String {
    for field in value.split(|c: char| !c.is_ascii_digit()) {
        if field.len() == 4 && field.starts_with("20") {
            return field.to_string();
        }
    }
    String::new()
}
